### ISSUE TITLE
fs.sshfs not threadsafe?

### ISSUE THREAD

**Issue by dhirschfeld** [althonos/fs.sshfs#17]
When I try to make multiple calls to `fs.open` in a `ThreadPoolExecutor` my program hangs :(

Is `fs.sshfs` expected to be threadsafe? Is this a known issue?

**Comment by althonos** [comment]
I expect it to be, since `paramiko.SFTPClient` is supposed to be as well. I tried to test against the new workers tests added by Will but unfortunately it always reaches a point where `SSHFile.read` hangs undefinitely, and I can't understand why yet.

**Comment by dhirschfeld** [comment]
Yeah, my app hung when I tried to use multiple threads. I've currently worked around it by using a threadlocal copy.

**Comment by abichinger** [comment]
> I expect it to be, since paramiko.SFTPClient is supposed to be as well.

I think paramiko is not thread safe
This [StackOverflow answer](https://stackoverflow.com/a/67636106/3140799) states that paramiko is not thread safe.
[Searching "thread" on paramiko.org](https://www.paramiko.org/search.html?q=thread) shows no results related to thread safty.

According to the [PyFilesystem documentation](https://docs.pyfilesystem.org/en/stable/implementers.html#thread-safety)

> All Filesystems should be thread-safe

I would appreciate it if fs.sshfs was also thread safe.
I'm going to look at it, maybe I can provide a solution.

Some things I have noticed so far:

- https://github.com/althonos/fs.sshfs/blob/020994638c19ec8d835bd1c3ef10d9ea8d165561/fs/sshfs/sshfs.py#L332
  I think the underlying channel never gets closed
- Making concurrent calls on `SSHFS._sftp` can cause a deadlock.
  For example, the following code snipped hangs on my system
  ```python
  sftp = client.open_sftp()
  
  def stat(path):
    return sftp.stat(path)

  with ThreadPoolExecutor(5) as e:
    futures = [e.submit(stat, "/home/root/.bashrc") for i in range(20)]
    for future in futures:
        res = future.result()
        print(res)
  ```
